R/log_shiny_input_changes.R

```diff
@@ -37,35 +38,38 @@ log_shiny_input_changes <- function(
     input,
     namespace = NA_character_,
     excluded_inputs = character(),
-    excluded_pattern = "_width$") {
-  session <- shiny::getDefaultReactiveDomain()
-  if (!(shiny::isRunning() || inherits(session, "MockShinySession") || inherits(session, "session_proxy"))) {
-    stop("No Shiny app running, it makes no sense to call this function outside of a Shiny app")
+    excluded_pattern = "_width$",
+    session = shiny::getDefaultReactiveDomain()) {
+  stopifnot(inherits(input, "reactivevalues"))
+  stopifnot(is.character(namespace) && length(namespace) == 1)
+  stopifnot(is.character(excluded_inputs))
+  stopifnot(is.character(excluded_pattern) && length(excluded_pattern) == 1)
+  stopifnot(inherits(session, "session_proxy"))
+
+  if (logger::TRACE > logger::as.loglevel(get_val("TEAL.LOG_LEVEL", "teal.log_level", "INFO"))) {
+    # to avoid setting observers when not needed
+    return(invisible(NULL))
   }
-  ns <- ifelse(!is.null(session), session$ns(character(0)), "")
 
-  # utils::assignInMyNamespace and utils::assignInNamespace are needed
-  # so that observer is executed only once, not twice.
-  input_values <- shiny::isolate(shiny::reactiveValuesToList(input))
-  utils::assignInMyNamespace("shiny_input_values", input_values)
+  ns <- session$ns(character(0))
+  reactive_input_list <- shiny::reactive({
+    input_list <- shiny::reactiveValuesToList(input)
+    input_list[!grepl(excluded_pattern, names(input_list))]
+  })
+  shiny_input_values <- shiny::reactiveVal(shiny::isolate(reactive_input_list()))
 
-  shiny::observe({
-    old_input_values <- shiny_input_values
-    new_input_values <- shiny::reactiveValuesToList(input)
-    names <- unique(c(names(old_input_values), names(new_input_values)))
-    names <- setdiff(names, excluded_inputs)
-    if (length(excluded_pattern)) {
-      names <- grep(excluded_pattern, names, invert = TRUE, value = TRUE)
-    }
+  shiny::observeEvent(reactive_input_list(), {
+    old_input_values <- shiny_input_values()
+    new_input_values <- reactive_input_list()
+    names <- intersect(names(old_input_values), names(new_input_values))
     for (name in names) {
-      old <- old_input_values[name]
-      new <- new_input_values[name]
+      old <- unname(old_input_values[name])
+      new <- unname(new_input_values[name])
       if (!identical(old, new)) {
         message <- trimws("{ns} Shiny input change detected in {name}: {old} -> {new}")
```

Comment:
- what do you think about introducing quoting?
```
r$> logger::log_info("Change in {name}: {old} -> {new}")
INFO [2024-09-05 12:12:32] Change in x: 1 -> aa

r$> logger::log_info("Change in {glue::backtick(name)}: {glue::backtick(old)} -> {
    glue::backtick(new)}")
INFO [2024-09-05 12:12:36] Change in `x`: `1` -> `aa`
```
- we must use collapsing mechanism to secure against lenghty `old` and `new` values
```
r$> logger::log_info("Change in {name}: {1:3} -> {1:2}")
Error in value[[3L]](cond) : `glue` failed in `formatter_glue` on:

  chr "Change in {name}: {1:3} -> {1:2}" 

Raw error message:

 Variables must be length 1 or 3 

Please consider using another `log_formatter` or `skip_formatter` on strings with curly braces.

r$> logger::log_info("Change in {name}: {glue::glue_collapse(1:3, sep = \", \")} -
    > {glue::glue_collapse(1:2, sep = \", \")}")
INFO [2024-09-05 12:15:51] Change in x: 1, 2, 3 -> 1, 2
```

Reply:
quoting handled already by `dput`. When character then it adds a double-qoute. `dput` doesn't put values/symbols/calls in the ticks.